src/name_core.jl

```diff
@@ -268,23 +268,40 @@ is_noninteger_type(::Any) = true
 
 """
     remaining_dimnames_from_indexing(dimnames::Tuple, inds)
-Given a tuple of dimension names
-and a set of index expressesion e.g `1, :, 1:3, [true, false]`,
-determine which are not dropped.
-Dimensions indexed with scalars are dropped
+
+Given a tuple of dimension names,
+and a tuple of indices e.g `(1, :, 1:3, [true, false])`,
+this drops those indexed with scalars or `CartesianIndex`,
+inserts `:_` for `newaxis = [CartesianIndex{0}()]`,
+and returns another tuple of names.
+
+It may also return `nothing`, to indicate that names should be dropped.
+This happens e.g. when indexing a matrix by a `BitArray{2}` such as `mat[mat .> 0.5]`,
+or by a vector of `CartesianIndex`es.
 """
-@generated function remaining_dimnames_from_indexing(dimnames::Tuple, inds)
+@generated function remaining_dimnames_from_indexing(dimnames::Tuple, inds::Tuple)
     # 0-Allocation see:
     # `@btime (()->remaining_dimnames_from_indexing((:a, :b, :c), (:,390,:)))()``
+    if length(inds.parameters) == 1 && length(dimnames.parameters) >= 2
+        type = inds.parameters[1]
+        if type <: BitArray || type <: AbstractVector{<:CartesianIndex}
+            return nothing
+        end
+    end
     keep_names = []
     dim_num = 0
     for type in inds.parameters
         if type <: Integer
```

Comment:
Lets make a functiion that is called inside each step of this loop:
that dispatches on `type`,
and does everything? Or returns somethat that is used to do everything?

Reply:
So it returns `dim_increment, new_names` or something? I don't think you can dispatch on `ndims(type) > 1`, so there will be some duplication. Not sure.

Reply:
`ndims(type)` you basically can for Tuples, and for arrays you dispatch on `AbstractVector` and then assume all others are >1.
(optionally also handle the `AbstractArray{T,0}`)

Reply:
Yea that's what I meant by duplication, it would need multiple paths to the "fall-through" case `else` here. All I've changed here is adding one more case to this list, am not so keen to re-write the rest.

Reply:
If we were to re-write this function, I suspect it should be done by recursively peeling elements off these two tuples. It probably need not be generated. But I have not done this.